contracts/zero-ex/contracts/src/transformers/FillOrdersTransformer.sol

```diff
@@ -0,0 +1,436 @@
+        // `sellToken`.
+        uint256 fillAmount;
+    }
+
+    /// @dev Results of a call to `_fillOrder()`.
+    struct FillOrderResults {
+        // The amount of taker tokens sold, according to balance checks.
+        uint256 takerTokenSoldAmount;
+        // The amount of maker tokens sold, according to balance checks.
+        uint256 makerTokenBoughtAmount;
+        // The amount of protocol fee paid.
+        uint256 protocolFeePaid;
+    }
+
+    /// @dev The Exchange ERC20Proxy ID.
+    bytes4 private constant ERC20_ASSET_PROXY_ID = 0xf47261b0;
+    /// @dev Maximum uint256 value.
+    uint256 private constant MAX_UINT256 = uint256(-1);
+
+    /// @dev The Exchange contract.
+    IExchange public immutable exchange;
+    /// @dev The ERC20Proxy address.
+    address public immutable erc20Proxy;
+
+    /// @dev Create this contract.
+    /// @param exchange_ The Exchange V3 instance.
+    constructor(IExchange exchange_)
+        public
+        Transformer()
+    {
+        exchange = exchange_;
+        erc20Proxy = exchange_.getAssetProxy(ERC20_ASSET_PROXY_ID);
+    }
+
+    /// @dev Sell this contract's entire balance of of `sellToken` in exchange
+    ///      for `buyToken` by filling `orders`. Protocol fees should be attached
+    ///      to this call. `buyToken` and excess ETH will be transferred back to the caller.
+    /// @param data_ ABI-encoded `TransformData`.
+    /// @return success The success bytes (`LibERC20Transformer.TRANSFORMER_SUCCESS`).
+    function transform(
+        bytes32, // callDataHash,
+        address payable, // taker,
+        bytes calldata data_
+    )
+        external
+        override
+        returns (bytes4 success)
+    {
+        TransformData memory data = abi.decode(data_, (TransformData));
+
+        // Validate data fields.
+        if (data.sellToken.isTokenETH() || data.buyToken.isTokenETH()) {
+            LibTransformERC20RichErrors.InvalidTransformDataError(
+                LibTransformERC20RichErrors.InvalidTransformDataErrorCode.INVALID_TOKENS,
+                data_
+            ).rrevert();
+        }
+        if (data.orders.length != data.signatures.length) {
+            LibTransformERC20RichErrors.InvalidTransformDataError(
+                LibTransformERC20RichErrors.InvalidTransformDataErrorCode.INVALID_ARRAY_LENGTH,
+                data_
+            ).rrevert();
+        }
+
+        if (data.side == Side.Sell && data.fillAmount == MAX_UINT256) {
+            // If `sellAmount == -1 then we are selling
+            // the entire balance of `sellToken`. This is useful in cases where
+            // the exact sell amount is not exactly known in advance, like when
+            // unwrapping Chai/cUSDC/cDAI.
+            data.fillAmount = data.sellToken.getTokenBalanceOf(address(this));
+        }
+
+        // Approve the ERC20 proxy to spend `sellToken`.
+        data.sellToken.approveIfBelow(erc20Proxy, data.fillAmount);
```

Comment:
Approval will be wrong for buys because `data.fillAmount` is in the maker asset. Will need to reflect this in the other transformer too.
```suggestion
 {
            uint256 initialSellTokenBalance =
                data.sellToken.getTokenBalanceOf(address(this));
            if (data.side == Side.Sell && data.fillAmount == MAX_UINT256) {
                // If `sellAmount == -1 then we are selling
                // the entire balance of `sellToken`. This is useful in cases where
                // the exact sell amount is not exactly known in advance, like when
                // unwrapping Chai/cUSDC/cDAI.
                data.fillAmount = initialSellTokenBalance;
            }
            // Approve the ERC20 proxy to spend `sellToken`.
            data.sellToken.approveIfBelow(erc20Proxy, initialSellTokenBalance);
        }
```

Reply:
As discussed this shouldn't have an impact at the approval is set to UINT256.

Reply:
I think we still need to make this change in case another transformer doesn't set the approval to unlimited.